azure/services/dns/arecords.go

```diff
@@ -2,83 +2,137 @@ package dns
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-08-01/network" //nolint
 	"github.com/Azure/go-autorest/autorest/to"
+
 	"github.com/giantswarm/microerror"
+	"github.com/go-logr/logr"
 	capzazure "sigs.k8s.io/cluster-api-provider-azure/azure"
+	capzpublicips "sigs.k8s.io/cluster-api-provider-azure/azure/services/publicips"
+	"sigs.k8s.io/controller-runtime/pkg/log"
 
 	"github.com/giantswarm/dns-operator-azure/azure"
 )
 
-func (s *Service) createARecords(ctx context.Context, aRecords []azure.ARecordSetSpec) error {
-	var err error
-	dnsSpec := s.Scope.DNSSpec()
-	if len(aRecords) == 0 {
-		s.Scope.V(2).Info(
+func (s *Service) calculateMissingARecords(ctx context.Context, logger logr.Logger, currentRecordSets []*armdns.RecordSet) []azure.ARecordSetSpec {
+	desiredRecordSet := s.getDesiredARecords()
+
+	var aRecordsToCreate []azure.ARecordSetSpec
+
+	for _, desiredARecordSet := range desiredRecordSet {
+		for _, recordSet := range currentRecordSets {
+			if recordSet.Type != nil && *recordSet.Type == RecordSetTypeA &&
+				recordSet.Name != nil && *recordSet.Name == desiredARecordSet.Hostname {
+				logger.Info(
+					fmt.Sprintf("DNS A record '%s' found", desiredARecordSet.Hostname),
+					"DNSZone", s.scope.ClusterDomain(),
+					"hostname", desiredARecordSet.Hostname,
+				)
+				continue
+			}
+
+			aRecordsToCreate = append(aRecordsToCreate, desiredARecordSet)
+			logger.Info(
+				fmt.Sprintf("DNS A record '%s' is missing, it will be created", desiredARecordSet.Hostname),
+				"DNSZone", "",
+				"hostname", desiredARecordSet.Hostname)
+		}
+
+	}
+
+	return aRecordsToCreate
+
+}
+
+func (s *Service) updateARecords(ctx context.Context, currentRecordSets []*armdns.RecordSet) error {
+	logger := log.FromContext(ctx).WithName("arecords")
+	recordsToCreate := s.calculateMissingARecords(ctx, logger, currentRecordSets)
+
+	zoneName := s.scope.ClusterZoneName()
+
+	if len(recordsToCreate) == 0 {
+		logger.Info(
 			"All DNS A records have already been created",
-			"DNSZone", dnsSpec.ZoneName)
+			"DNSZone", zoneName)
 		return nil
 	}
 
-	for _, aRecord := range aRecords {
-		var ipAddressObject network.PublicIPAddress
-		ipAddressObject, err = s.publicIPsService.Get(ctx, s.Scope.ResourceGroup(), aRecord.PublicIPName)
+	for _, aRecord := range recordsToCreate {
+		ipAddressObject, err := s.publicIPsService.Get(ctx, &capzpublicips.PublicIPSpec{
+			Name:          aRecord.PublicIPName,
+			ResourceGroup: s.scope.ResourceGroup(),
+		})
 		if capzazure.ResourceNotFound(err) {
-			s.Scope.V(2).Info(
+			logger.Info(
 				"Cannot create DNS A record, public IP still not deployed",
-				"DNSZone", dnsSpec.ZoneName,
+				"DNSZone", zoneName,
 				"hostname", aRecord.Hostname,
 				"IP resource name", aRecord.PublicIPName)
 			continue
 		} else if err != nil {
 			return microerror.Mask(err)
 		}
 
-		if ipAddressObject.IPAddress == nil {
-			s.Scope.V(2).Info(
+		ipAddress := ipAddressObject.(network.PublicIPAddress).IPAddress
```

Comment:
This is the reason why we need the`github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-08-01/network` import. Since this is deprecated, our linter fails and had to silence it. Not sure if there is a better workaround now though since `publicIPsService.Get` still returns the old type.